### PR TITLE
37 clear button

### LINKED ISSUE #36: Plot first scan automatically
Double clicking on the first graph plots the default detector vs positioner.

Subsequent selections depend on the selection in #35

### LINKED ISSUE #43: Implement Matplotlib panel
pyQtgraph is not intuitive to interact with, probably better for images.

### LINKED ISSUE #44: Add Clear graph button
Not sure we want Remove button to do that. A different button might be better.

### LINKED ISSUE #46: Implement cursors
Add cursors + info panel for cursors position and basic math m(difference, midpoint)

### LINKED ISSUE #51: app crashes when trying to open folder with wrong permissions: drwx------
```bash
Traceback (most recent call last):
  File "/home/beams22/29IDUSER/bin/mdaviz/mdaviz/mainwindow.py", line 177, in setFolderPath
    self.setSubfolderList(get_all_subfolders(folder_path, folder_path.name))
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/beams22/29IDUSER/bin/mdaviz/mdaviz/mainwindow.py", line 174, in get_all_subfolders
    subfolder_list += get_all_subfolders(item, new_parent_path)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/beams22/29IDUSER/bin/mdaviz/mdaviz/mainwindow.py", line 174, in get_all_subfolders
    subfolder_list += get_all_subfolders(item, new_parent_path)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/beams22/29IDUSER/bin/mdaviz/mdaviz/mainwindow.py", line 165, in get_all_subfolders
    for item in folder_path.iterdir():
  File "/home/beams/29IDUSER/.conda/envs/mdaviz/lib/python3.11/pathlib.py", line 931, in iterdir
    for name in os.listdir(self):
                ^^^^^^^^^^^^^^^^
PermissionError: [Errno 13] Permission denied: '/net/s29data/export/old_data_29idd/lost+found'
^C
[1]+  Aborted                 (core dumped) mdaviz
```

### PR DESCRIPTION
**Pull request by rodolakis**
- close #36 
- close #43
- close #44 
- close #46
- close #51
- #52

### CODE REVIEW

File: env.yml [hunk 1]
```diff
@@ -12,6 +12,7 @@ dependencies:
   - pyqtgraph
   - pyRestTable
   - qt =5
+  - matplotlib
```
Comment (prjemian) [inline env.yml:15]: sort (within section)

File: mdaviz/chartview.py [hunk 2]
```diff
@@ -117,3 +123,203 @@ def setLeftAxisUnits(self, text):
...
+        self.main_axes.set_xlabel(text, fontsize=FONTSIZE, labelpad=10)
+
+    def setLeftAxisText(self, text):
+        self.main_axes.set_ylabel(text, fontsize=FONTSIZE, labelpad=20)
+
+    def clearPlot(self):
+        self.main_axes.clear()
+        self.cursors[1] = False
+        self.cursors[2] = False
+        self.info_panel_axes.clear()
```
Comment (prjemian) [inline mdaviz/chartview.py:208]:
```suggestion
        self.cursors["pos1"] = None
        self.cursors["pos2"] = None
```

File: mdaviz/chartview.py [hunk 3]
```diff
@@ -117,3 +123,203 @@ def setLeftAxisUnits(self, text):
...
+    # TODO: use a dictionary to track the cursors
+    # TODO: allow a blue cursor to be placed without the presence of a red cursor
+
+    def onclick(self, event):
+        # Check if the click was in the main_axes
+        if event.inaxes is self.main_axes:
+            # Middle click for red cursor
+            if event.button == 2:
+                if self.cursors[1]:
+                    self.cursors[1].remove()  # Remove existing red cursor
```
Comment (prjemian) [inline mdaviz/chartview.py:222]: If `self.cursors[1]` is a boolean, how does it have a `remove()` method?

Comment (prjemian) [reply mdaviz/chartview.py:222]: It's either `False` (default) or an artist structure from MPL.  Might be better to default as `None` (which implies some logic changes that test for `None`).

File: mdaviz/chartview.py [hunk 4]
```diff
@@ -117,3 +123,203 @@ def setLeftAxisUnits(self, text):
...
+                (self.cursors[1],) = self.main_axes.plot(
+                    event.xdata, event.ydata, "r+", markersize=15, linewidth=2
+                )
+                # Update cursor position
+                self.cursors["pos1"] = (event.xdata, event.ydata)
+
+            # Right click for blue cursor
+            elif event.button == 3:
+                if self.cursors[2]:
+                    self.cursors[2].remove()  # Remove existing blue cursor
```
Comment (prjemian) [inline mdaviz/chartview.py:232]: same thing

File: mdaviz/chartview.py [hunk 5]
```diff
@@ -117,3 +123,203 @@ def setLeftAxisUnits(self, text):
...
+                self.cursors["pos2"] = (event.xdata, event.ydata)
+
+            # Update the info panel with cursor positions
+            self.update_info_panel()
+
+            # Redraw the canvas to display the new markers
+            self.canvas.draw()
+
+    def update_info_panel(self):
+        # Clear the info panel by removing each text artist
```
Comment (prjemian) [inline mdaviz/chartview.py:247]: convert to docstring

File: mdaviz/chartview.py [hunk 6]
```diff
@@ -117,3 +123,203 @@ def setLeftAxisUnits(self, text):
...
+        while len(self.info_panel_axes.texts) > 0:
+            for text in self.info_panel_axes.texts:
+                text.remove()
+
+        # Initialize text variables
+        cursor1_text = "Cursor 1: press middle click"
```
Comment (prjemian) [inline mdaviz/chartview.py:253]: Could this be new keys (`text1`, `text2`) in `self.cursors` dict?

File: mdaviz/chartview.py [hunk 7]
```diff
@@ -117,3 +123,203 @@ def setLeftAxisUnits(self, text):
...
+        cursor2_text = "Cursor 2: press right click"
+        diff_text = "Difference: n/a"
```
Comment (prjemian) [inline mdaviz/chartview.py:255]: Similar, these keys could be in a dictionary that contains these computed results.

File: mdaviz/mda_file_viz.py [hunk 8]
```diff
@@ -32,11 +34,37 @@ def setMetadata(self, text, *args, **kwargs):
...
         layout.addWidget(plot_widget)
-        self.tabWidget.setCurrentWidget(self.plotPage)
+        self.tabWidget.setCurrentWidget(self.plotPageQt)
+
+    def setPlotMpl(self, plot_widget):
+        layout = self.plotPageMpl.layout()
+        utils.removeAllLayoutWidgets(layout)
+        layout.addWidget(plot_widget)
+        self.tabWidget.setCurrentWidget(self.plotPageMpl)
+
```
Comment (prjemian) [inline mdaviz/mda_file_viz.py:48]:
```suggestion

    def isPlotBlank(self, plotPage, chartView):
        layout = plotPage.layout()
        if layout.count() == 0:
            return True
        plot_widget = layout.itemAt(0).widget()
        # Check if the plot widget is an instance of chartView and has data items
        if isinstance(plot_widget, chartView):
            return not plot_widget.hasDataItems()
        return True  # If not a chartView instance, consider it as blank

    def isPlotBlankQt(self):
        return self.isPlotBlank(self.plotPageQt, ChartViewQt)

    def isPlotBlankMpl(self):
        return self.isPlotBlank(self.plotPageMpl, ChartViewMpl)
```

Comment (prjemian) [reply mdaviz/mda_file_viz.py:48]: Similar for `setPlotQt` and `setPlotMpl`

File: mdaviz/mda_folder.py [hunk 9]
```diff
@@ -76,6 +76,16 @@ def doFileSelected(self, index):
             self.select_fields_tableview.selected.connect(self.doPlot)
             self.setStatus(f"Selected file: {self.mdaFileList()[index.row()]}")
 
+            if self.mda_file_visualization.isPlotBlankQt():  # TODO what about MPL?
```
Comment (prjemian) [inline mdaviz/mda_folder.py:79]: What about this `TODO` marker?

File: mdaviz/mda_folder.py [hunk 10]
```diff
@@ -76,6 +76,16 @@ def doFileSelected(self, index):
...
+                # TODO: this should depend on the selection: auto-replace vs auto-add
+                first_pos_idx = self.select_fields_tableview.firstPos()
+                first_det_idx = self.select_fields_tableview.firstDet()
+                if first_pos_idx is not None and first_det_idx is not None:
+                    first_selections = {"X": first_pos_idx, "Y": [first_det_idx]}
+                    self.doPlot("replace", first_selections)
+                else:
+                    self.setStatus("Could not find a positioner or detector to plot.")
```
Comment (prjemian) [inline mdaviz/mda_folder.py:87]:
```suggestion
                    self.setStatus("Could not find a positioner and detector to plot.")
```

File: mdaviz/mda_folder.py [hunk 11]
```diff
@@ -76,6 +76,16 @@ def doFileSelected(self, index):
             self.select_fields_tableview.selected.connect(self.doPlot)
             self.setStatus(f"Selected file: {self.mdaFileList()[index.row()]}")
 
+            if self.mda_file_visualization.isPlotBlankQt():  # TODO what about MPL?
+                # TODO: this should depend on the selection: auto-replace vs auto-add
+                first_pos_idx = self.select_fields_tableview.firstPos()
+                first_det_idx = self.select_fields_tableview.firstDet()
+                if first_pos_idx is not None and first_det_idx is not None:
+                    first_selections = {"X": first_pos_idx, "Y": [first_det_idx]}
+                    self.doPlot("replace", first_selections)
```
Comment (prjemian) [inline mdaviz/mda_folder.py:85]: Does `doPlot()` handle question of Qt or Mpl?  Can it relieve the above TODO marker?

File: mdaviz/mda_folder.py [hunk 12]
```diff
@@ -110,34 +120,56 @@ def doRefresh(self):
...
-        from .chartview import ChartView
-        from .select_fields_table_view import to_datasets
+        from .chartview import ChartViewQt
+        from .chartview import ChartViewMpl
+        from .select_fields_table_view import to_datasets_qt
+        from .select_fields_table_view import to_datasets_mpl
 
         action = args[0]
         selections = args[1]
+        print(f"doPlot called with action: {action}, args: {args}")
```
Comment (prjemian) [inline mdaviz/mda_folder.py:130]: Ebentually, refactor as logger.debug instead of print

File: mdaviz/mda_folder.py [hunk 13]
```diff
@@ -110,34 +120,56 @@ def doRefresh(self):
...
 
         detsDict = self.select_fields_tableview.detsDict()
 
         # setup datasets
-        datasets, options = to_datasets(detsDict, selections)
+        datasets_qt, options_qt = to_datasets_qt(detsDict, selections)
+        datasets_mpl, options_mpl = to_datasets_mpl(detsDict, selections)
+
+        # # get the pyQtchart chartview widget, if exists:
```
Comment (prjemian) [inline mdaviz/mda_folder.py:138]: remove extra `#`

File: mdaviz/mda_folder.py [hunk 14]
```diff
@@ -110,34 +120,56 @@ def doRefresh(self):
...
+            widgetQt.clearPlot()
+            widgetMpl.clearPlot()
 
         if action in ("replace", "add"):
-            for ds, ds_options in datasets:
-                widget.plot(*ds, **ds_options)
-            self.mda_file_visualization.setPlot(widget)
+            for ds, ds_options in datasets_qt:
+                widgetQt.plot(*ds, **ds_options)
+            self.mda_file_visualization.setPlotQt(widgetQt)
```
Comment (prjemian) [inline mdaviz/mda_folder.py:169]: Refactor for common `setPlot` method

File: mdaviz/mda_folder.py (hunk 11 above)
Comment (rodolakis) [reply mdaviz/mda_folder.py:85]: create a new issue #55

File: mdaviz/mda_folder.py (hunk 12 above)
Comment (rodolakis) [reply mdaviz/mda_folder.py:130]: created a new issue #56

File: mdaviz/mda_folder.py (hunk 14 above)
Comment (rodolakis) [reply mdaviz/mda_folder.py:169]: added to #55

File: mdaviz/mda_folder.py (hunk 9 above)
Comment (rodolakis) [reply mdaviz/mda_folder.py]: At the moment both MPL and Qt are handle identically (clear clears both graphs, add adds to both graph etc...) so if MPL is blank, Qt is blank.

**Review by prjemian — APPROVED** [review]
LGTM!  Nice!

**Pull request merged**